cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java

```diff
@@ -122,4 +253,372 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
     return String.format("%s:%s:%s:%s", program.getNamespaceId(), program.getApplicationId(),
                          programType.name(), program.getId());
   }
+
+  /**
+   * One instance of this class contains a list of {@link StreamSizeSchedule}s, which are all interested
+   * in the same stream. This instance subscribes to the size notification of the stream, and polls the
+   * stream for its size whenever the schedules it references need the information.
+   * The {@link StreamSizeScheduler} communicates with this class, which in turn communicates to the schedules
+   * it contains to perform operations on the schedules - suspend, resume, etc.
+   */
+  private final class StreamSubscriber implements NotificationHandler<StreamSizeNotification>, Cancellable {
+    // Key is the schedule ID
+    private final ConcurrentMap<String, StreamSizeScheduleTask> scheduleTasks;
+    private final Object lastNotificationLock;
+    private final Id.Stream streamId;
+    private final AtomicInteger activeTasks;
+
+    private Cancellable notificationSubscription;
+    private ScheduledFuture<?> scheduledPolling;
+    private StreamSizeNotification lastNotification;
+
+    private StreamSubscriber(Id.Stream streamId) {
+      this.streamId = streamId;
+      this.scheduleTasks = Maps.newConcurrentMap();
+      this.lastNotificationLock = new Object();
+      this.activeTasks = new AtomicInteger(0);
+    }
+
+    public void start() throws NotificationFeedException, NotificationFeedNotFoundException {
```

Comment:
Also, there could be some inconsistency in the app deployment phase. So far I think that we considered that creating a schedule could never fail, and so there is no exception handling when creating the schedules, when deploying an app. But now, this new schedule can fail if the stream it references doesn't exist. 

Yet, with the current behavior, 1. if one schedule fails to be created, then we don't even try to create the remaining schedules, and 2. the app is still deployed, and the ApplicationSpecification references a schedule which could not be started. What should be done here? 

If we don't want to make the app deployment fail when a stream doesn't exist, then I think we will have to wait until that JIRA - https://issues.cask.co/browse/CDAP-1446 - gets resolved. Basically, it means that the schedule would be created, but not started. The schedule should wait till the stream is created to start.